### PR TITLE
feat: replace @napi-rs/pinyin by pinyin-pro

### PR DESCRIPTION
**Pull request by starcsb**

### CODE REVIEW

**Comment by starcsb** [comment]
close #47

**Comment by Brooooooklyn** [comment]
This pr will decrease the performance while slugging Chinese input, ref: https://github.com/zh-lx/pinyin-pro/issues/99

**Comment by starcsb** [comment]
> This pr will decrease the performance while slugging Chinese input, ref: [zh-lx/pinyin-pro#99](https://github.com/zh-lx/pinyin-pro/issues/99)
> 
> And I think the accurate issue is really easy to solve; the https://github.com/mozillazg/pinyin-data hasn't been updated for a long time.

I don't think it will decrease the performance. By my testing, only when the length of the text exceeds about `300k` can `@napi-rs/pinyin` perform better than `pinyin-pro`, and the performance difference does not exceed twice. But for most users, their input text will not too long. When the text length is within 100, the translation time for `pinyin-pro` is only 1ms, while the translation time for `@napi-rs/pinyin` is 150 milliseconds. Therefore, in the vast majority of cases, the performance of `pinyin-pro` is much higher than that of `@napi-rs/pinyin`.

**Comment by Brooooooklyn** [comment]
As my description in https://github.com/zh-lx/pinyin-pro/issues/99, you are comparing `pinyin-pro` with the `segment` option enabled with `@napi-rs/pinyin`, which is unfair; because 1. `pinyin-pro` doesn't support `segment` feature 2. `limax` project doesn't use `segment` feature. From my benchmark, for the **100 words input**, the benchmark result shows:

```
❯ node speed.js  
pinyin-pro 转换 100 字数时间: 0.582ms
pinyin 转换 100 字数时间: 0.292ms
@napi-rs/pinyin 转换 100 字数时间: 0.041ms
```

Which is significantly different

**Comment by starcsb** [comment]
> As my description in [zh-lx/pinyin-pro#99](https://github.com/zh-lx/pinyin-pro/issues/99), you are comparing `pinyin-pro` with the `segment` option enabled with `@napi-rs/pinyin`, which is unfair; because 1. `pinyin-pro` doesn't support `segment` feature 2. `limax` project doesn't use `segment` feature. From my benchmark, for the **100 words input**, the benchmark result shows:
> 
> ```
> ❯ node speed.js  
> pinyin-pro 转换 100 字数时间: 0.582ms
> pinyin 转换 100 字数时间: 0.292ms
> @napi-rs/pinyin 转换 100 字数时间: 0.041ms
> ```
> 
> Which is significantly different

As the author's reply, pinyin-pro defaults to having segment enabled in all cases(otherwise it wouldn't have such high accuracy), this option doesn't need to be opened by the user manually. So the performance should by compared with `segment` enabled.

**Comment by Brooooooklyn** [comment]
> So the performance should by compared with segment enabled

Okay, sure, but the `limax` doesn't need the `segment` option, the `@napi-rs/pinyin` is still ***14x faster*** in this case.

**Comment by starcsb** [comment]
> > So the performance should by compared with segment enabled
> 
> Okay, sure, but the `limax` doesn't need the `segment` option, the `@napi-rs/pinyin` is still _**14x faster**_ in this case.

But if don't open `segment`, how to ensure the accuracy? I think `segment` is the key of improving accuracy.

**Comment by starcsb** [comment]
The core purpose of this PR is to improve accuracy rather than performance. The upgrade of [mozillazg/pinyin-data](https://github.com/mozillazg/pinyin-data) does not significantly improve accuracy

**Comment by Brooooooklyn** [comment]
@starcsb you are right, after checking the source code from `pinyin-pro`, I realized that combing segment and phrase is a better way to improve both performance and accuracy. Thanks for the discussion, I'll post these back to the `rust-pinyin`.

File: package.json [hunk 1]
```diff
@@ -52,7 +52,7 @@
     "speakingurl": "^14.0.1"
   },
   "optionalDependencies": {
-    "@napi-rs/pinyin": "^1.7.1"
+    "pinyin-pro": "^3.14.0"
```
Comment (lovell) [inline package.json:55]: Given `pinyin-pro` is not a native module, we could promote it to standard `dependencies` and remove the try/catch block around its `require` statement.

File: test/unit.js [hunk 2]
```diff
@@ -71,6 +71,12 @@ ava('Set Pinyin tone numbering via tone option', function (t) {
   t.true(
     limax('特殊天-1', { tone: false }) === 'te-shu-tian-1'
   );
+  t.true(
```
Comment (lovell) [inline test/unit.js:74]: Thanks for adding tests. These will also require the following change:
```diff
- t.plan(3);
+ t.plan(5);
```

**Comment by starcsb** [comment]
> Thank you for the PR, I've left a couple of comments inline.

Thanks for your comments, I have modified these.

**Comment by starcsb** [comment]
> Thank you for the PR, I've left a couple of comments inline.

Thanks for your comments, I have modified them.

File: lib/limax.js [hunk 3]
```diff
@@ -4,9 +4,7 @@ const speakingurl = require('speakingurl');
 const hepburn = require('hepburn');
 const deburr = require('lodash.deburr');
 let pinyin, PINYIN_STYLE;
```
Comment (lovell) [inline lib/limax.js:6]: Line 6 can be removed.

**Review by lovell — CHANGES_REQUESTED** [review]

**Comment by starcsb** [comment]
图片

@lovell I corrected a [original test case](https://github.com/lovell/limax/pull/48/files#diff-6c4ac77163e8ec13146b6400fa4efab193dab7668284713a1165867c62fcbbbbL14-R14). In chinese, the pinyin of `弄堂` should be `long4 tang2` rather than `nong4 tang2`. It still relies on `pinyin-pro` to discover this problem.

See more detail: [弄堂](https://baike.baidu.com/link?url=yL3DKJp2jVUoR0BC3iJds9gfDlJsSnqq4Lpr8y5JiauNO6mZ_3PL03JipDGjaGhh8GTKOpVoTAmD6uQCTEPF5S-ht3TfyeRi0MzO7hvtuNW)

图片

**Review by lovell — APPROVED** [review]

**Comment by lovell** [comment]
v4.1.0 now available, thank you.

**Pull request merged**